### PR TITLE
API Prototype

### PR DESCRIPTION
**Pull request by periodic**
This is a draft of what the API effect might look like, including tests.

### CODE REVIEW

File: src/Control/Carrier/FossaAPIClientIO.hs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+-- | Runs FossaAPI effects as IO operations
+-- TODO: Put ApiOpts in a reader
+-- TODO: Exception handling
+runFossaAPIClientIO ::
+  ( Has (Lift IO) sig m
+  , Has Diagnostics sig m
+  , Has Logger sig m
+  ) =>
+  FossaAPIClientIOC m a ->
+  m a
```
Comment (scruffystuffs) [inline src/Control/Carrier/FossaAPIClientIO.hs:24]: I think some later work to do is split out an HTTP effect from this, but for now this works fine.

Comment (scruffystuffs) [reply src/Control/Carrier/FossaAPIClientIO.hs:24]: And `runFossaApiClientIO` might have a prime version that also terminates the HTTP effect.

File: test/Test/Effect.hs [hunk 2]
```diff
@@ -109,3 +110,11 @@ shouldMatchList' a b = sendIO $ shouldMatchList a b
 
 expectFailure' :: Has (Lift IO) sig m => Result a -> m ()
 expectFailure' res = sendIO $ expectFailure res
+
+-- TODO: Better naming.
+expectFails' :: (Has Diagnostics sig m, Has (Lift IO) sig m) => m () -> m ()
+expectFails' f = do
```
Comment (scruffystuffs) [inline test/Test/Effect.hs:116]: Maybe `expectFatal`, since `fatal` is what `recover` catches.

File: src/Control/Carrier/FossaAPIClientIO.hs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+type FossaAPIClientIOC = SimpleC FossaAPIClientF
+
+-- | Runs FossaAPI effects as IO operations
+-- TODO: Put ApiOpts in a reader
+-- TODO: Exception handling
+runFossaAPIClientIO ::
+  ( Has (Lift IO) sig m
+  , Has Diagnostics sig m
+  , Has Logger sig m
+  ) =>
```
Comment (scruffystuffs) [inline src/Control/Carrier/FossaAPIClientIO.hs:22]: Let's add the `ReaderC ApiOpts` here now, and see if we can improve the testing api from that.

Comment (scruffystuffs) [reply src/Control/Carrier/FossaAPIClientIO.hs:22]: I think with some work, we can make `withMockApi` a lot easier to use, and more powerful.

**Comment by periodic** [comment]
Closing in favor of #843

**Pull request closed without merging**